revocation/crl/cache/memory.go

```diff
@@ -0,0 +1,76 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// MemoryCache is an in-memory cache that stores CRL bundles.
+//
+// The cache is built on top of the sync.Map to leverage the concurrency control
+// and atomicity of the map, so it is suitable for writing once and reading many
+// times. The CRL is stored in memory as a Bundle type.
+//
+// MemoryCache doesn't handle cache cleaning but provides the Delete and Clear
+// methods to remove the CRLs from the memory.
+type MemoryCache struct {
+	// MaxAge is the maximum age of the CRLs cache. If the CRL is older than
+	// MaxAge, it will be considered as expired.
+	MaxAge time.Duration
```

Comment:
Shouldnt we derive this value from CRL itself as it might vary among certificates
?

Reply:
According to 
> [5.1.2.5](https://www.rfc-editor.org/rfc/rfc5280#section-5.1.2.5).  Next Update
   This field indicates the date by which the next CRL will be issued.
   The next CRL could be issued before the indicated date, but it will
   not be issued any later than the indicated date.

The new CRL may be published before the next update. Also, the next update is an optional extension in the CRL. We need a default one to force the update.

Reply:
Can we make this mandatory as per CAB forum guideline?
https://cabforum.org/working-groups/server/baseline-requirements/documents/TLSBRv2.0.4.pdf

Reply:
I see. We can do that, but for those CAs that don’t follow the CAB Forum, we don’t provide any warranty.

Reply:
Here is the statistic:
```
processing CRL: http://crl3.digicert.com/CloudflareIncECCCA-3.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/CloudflareIncECCCA-3.crl
CRL next update duration: 7 days
processing CRL: http://crl.r2m01.amazontrust.com/r2m01.crl
CRL next update duration: 7 days
processing CRL: http://crl.r2m02.amazontrust.com/r2m02.crl
CRL next update duration: 7 days
processing CRL: http://crl.r2m03.amazontrust.com/r2m03.crl
CRL next update duration: 7 days
processing CRL: http://c.pki.goog/wr2/oBFYYahzgVI.crl
CRL next update duration: 9 days
processing CRL: http://c.pki.goog/wr2/9UVbN0w5E6Y.crl
CRL next update duration: 9 days
processing CRL: http://c.pki.goog/wr2/75r4ZyA3vA0.crl
CRL next update duration: 9 days
processing CRL: http://c.pki.goog/wr2/GSyT1N4PBrg.crl
CRL next update duration: 9 days
processing CRL: http://c.pki.goog/we1/ygWPENklvpM.crl
CRL next update duration: 9 days
processing CRL: http://c.pki.goog/we1/H9bdJBu1Tvg.crl
CRL next update duration: 9 days
processing CRL: http://c.pki.goog/we1/T58q3x0jyXI.crl
CRL next update duration: 9 days
processing CRL: http://c.pki.goog/we1/FARePxy0M2M.crl
CRL next update duration: 9 days
processing CRL: http://c.pki.goog/we1/OhMZjmT1BfY.crl
CRL next update duration: 9 days
processing CRL: http://crl3.digicert.com/sha2-ha-server-g6.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/sha2-ha-server-g6.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/DigiCertGlobalG2TLSRSASHA2562020CA1-1.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/DigiCertGlobalG2TLSRSASHA2562020CA1-1.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/DigiCertTLSHybridECCSHA3842020CA1-1.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/DigiCertTLSHybridECCSHA3842020CA1-1.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/DigiCertTLSRSASHA2562020CA1-4.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/DigiCertTLSRSASHA2562020CA1-4.crl
CRL next update duration: 7 days
processing CRL: http://cdp.rapidssl.com/RapidSSLECCCA2018.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/DigiCertGlobalCAG2.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/DigiCertGlobalCAG2.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/DigicertSHA2SecureServerCA-1.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/DigicertSHA2SecureServerCA-1.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/DigiCertSecureSiteECCCA-1.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/DigiCertSecureSiteECCCA-1.crl
CRL next update duration: 7 days
processing CRL: http://cdp.geotrust.com/GeoTrustTLSRSACAG1.crl
CRL next update duration: 7 days
processing CRL: http://cdp.geotrust.com/GeoTrustRSACA2018.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/sha2-ev-server-g3.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/sha2-ev-server-g3.crl
CRL next update duration: 7 days
processing CRL: http://cdp.thawte.com/ThawteRSACA2018.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/DigiCertGlobalG3TLSECCSHA3842020CA1-2.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/DigiCertGlobalG3TLSECCSHA3842020CA1-2.crl
CRL next update duration: 7 days
processing CRL: http://crl3.digicert.com/DigiCertEVRSACAG2.crl
CRL next update duration: 7 days
processing CRL: http://crl4.digicert.com/DigiCertEVRSACAG2.crl
CRL next update duration: 7 days
processing CRL: http://crl.sectigo.com/SectigoRSAOrganizationValidationSecureServerCA.crl
CRL next update duration: 7 days
processing CRL: http://crl.globalsign.com/ca/gsatlasr3dvtlsca2024q1.crl
CRL next update duration: 7 days
processing CRL: http://crl.globalsign.com/gsrsaovsslca2018.crl
CRL next update duration: 6 days
processing CRL: http://crl.globalsign.com/gsgccr3dvtlsca2020.crl
CRL next update duration: 6 days
processing CRL: http://crl.globalsign.com/ca/gsatlasr3dvtlsca2023q4.crl
CRL next update duration: 7 days
processing CRL: http://crl.globalsign.com/ca/gsatlasr3dvtlsca2024q2.crl
CRL next update duration: 7 days
processing CRL: http://crl.globalsign.com/ca/gsatlasr3dvtlsca2024q3.crl
CRL next update duration: 7 days
processing CRL: http://crl.globalsign.com/gseccovsslca2018.crl
CRL next update duration: 6 days
processing CRL: http://crl.globalsign.com/alphasslcasha256g4.crl
CRL next update duration: 6 days
processing CRL: http://crl.godaddy.com/gdig2s1-8229.crl
CRL next update duration: 7 days
processing CRL: http://crl.godaddy.com/gdig2s1-14080.crl
CRL next update duration: 7 days
processing CRL: http://www.microsoft.com/pkiops/crl/Microsoft%20Azure%20RSA%20TLS%20Issuing%20CA%2007.crl
CRL next update duration: 8 days
processing CRL: http://www.microsoft.com/pkiops/crl/Microsoft%20Azure%20RSA%20TLS%20Issuing%20CA%2008.crl
CRL next update duration: 8 days
processing CRL: http://www.microsoft.com/pkiops/crl/Microsoft%20Azure%20ECC%20TLS%20Issuing%20CA%2004.crl
CRL next update duration: 8 days
processing CRL: http://crl.apple.com/apevsrsa2g1.crl
CRL next update duration: 6 days
processing CRL: http://crl.entrust.net/level1m.crl
CRL next update duration: 6 days
processing CRL: http://crl.entrust.net/level1k.crl
CRL next update duration: 6 days
processing CRL: http://crl.comodoca.com/COMODOECCOrganizationValidationSecureServerCA.crl
CRL next update duration: 7 days
```

Reply:
Do you mean that we can remove `DefaultMaxAge`?